### PR TITLE
Add aliases to Dovecot's userdb & Adjust the passdb / userdb configuration

### LINKED ISSUE #2091: Weird Dovecot login failures for emails arriving to a virtual alias
# Subject

Understanding Dovecot login failures when email for an alias arrives. Email is working, just trying to understand the log messages.

## Description

I've finally taken the leap and moved my email server over to docker-mailserver. **Thank you to all of the contributors to this project.**

It's still early days for me - but I've run a postfix based mail server for years (just installed directly on the OS). I'm seeing something odd in the mail.log from my new installation of docker-mailserver.

An email comes in for `skincare@mydomain.com` - this is an alias in my `config/postfix-virtual.cf` file that is mapped to this user (casing is different, but that shouldn't matter)

Snippet of the virtual alias file
```
Skincare@mydomain.com      myuser@mydomain.com
```

Now if we look at the mail.log for that email arriving - we see a weird thing.
```
Jul 19 15:01:03 mail dovecot: auth: passwd-file(skincare@mydomain.com): unknown user 
```

Now the email does get delivered to the right account (the alias is working fine). My concern is about why dovecot is trying to log in as the alias user? What's up with that? I can turn on more debug - but this feels like a pretty plain deployment.

I'd like to avoid fail2ban catching too many alias emails arriving, and banning someone.

Here is the more complete log.
```
Jul 19 15:01:02 mail postfix/postscreen[28103]: CONNECT from [136.147.130.236]:57937 to [172.17.0.5]:25
Jul 19 15:01:02 mail postfix/dnsblog[28122]: addr 136.147.130.236 listed by domain list.dnswl.org as 127.0.15.0
Jul 19 15:01:02 mail postfix/postscreen[28103]: PASS NEW [136.147.130.236]:57937
Jul 19 15:01:03 mail postfix/smtpd[29759]: connect from mta.mail.arbonnemail.com[136.147.130.236]
Jul 19 15:01:03 mail postfix/smtpd[29759]: Anonymous TLS connection established from mta.mail.arbonnemail.com[136.147.130.236]: TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits)
Jul 19 15:01:03 mail policyd-spf[29769]: prepend Received-SPF: Pass (mailfrom) identity=mailfrom; client-ip=136.147.130.236; helo=mta.mail.arbonnemail.com; envelope-from=bounce-344_html-18444
363-1683099-7207065-3001@bounce.mail.arbonnemail.com; receiver=<UNKNOWN> 
Jul 19 15:01:03 mail dovecot: auth: passwd-file(skincare@mydomain.com): unknown user 
Jul 19 15:01:03 mail postfix/smtpd[29759]: CF302148566: client=mta.mail.arbonnemail.com[136.147.130.236]
Jul 19 15:01:03 mail postfix/cleanup[29773]: CF302148566: message-id=<69969198-6397-472c-ac73-609566daee5a@atl1s07mta2081.xt.local>
Jul 19 15:01:04 mail opendkim[2920]: CF302148566: mta.mail.arbonnemail.com [136.147.130.236] not internal
Jul 19 15:01:04 mail opendkim[2920]: CF302148566: not authenticated
Jul 19 15:01:04 mail opendkim[2920]: CF302148566: DKIM verification successful
Jul 19 15:01:04 mail opendkim[2920]: CF302148566: s=200608 d=mail.arbonnemail.com SSL 
Jul 19 15:01:04 mail opendmarc[2927]: CF302148566 ignoring Authentication-Results at 1 from mail.mydomain.com
Jul 19 15:01:04 mail opendmarc[2927]: CF302148566: mail.arbonnemail.com none
Jul 19 15:01:04 mail postfix/qmgr[3643]: CF302148566: from=<bounce-344_HTML-18444363-1683099-7207065-3001@bounce.mail.arbonnemail.com>, size=108763, nrcpt=1 (queue active)
Jul 19 15:01:04 mail postfix/smtpd[29759]: disconnect from mta.mail.arbonnemail.com[136.147.130.236] ehlo=2 starttls=1 mail=1 rcpt=1 data=1 quit=1 commands=7
Jul 19 15:01:09 mail postfix/smtpd[29808]: connect from localhost[127.0.0.1]
Jul 19 15:01:09 mail postfix/smtpd[29808]: DE6F2148568: client=localhost[127.0.0.1]
Jul 19 15:01:09 mail postfix/cleanup[29773]: DE6F2148568: message-id=<69969198-6397-472c-ac73-609566daee5a@atl1s07mta2081.xt.local>
Jul 19 15:01:09 mail postfix/smtpd[29808]: disconnect from localhost[127.0.0.1] ehlo=1 mail=1 rcpt=1 data=1 quit=1 commands=5
Jul 19 15:01:09 mail amavis[3650]: (03650-04) Passed CLEAN {RelayedInbound}, [136.147.130.236]:57937 [136.147.130.236] <bounce-344_HTML-18444363-1683099-7207065-3001@bounce.mail.arbonnemail.com> -> <myuser@mydomain.com>, Queue-ID: CF302148566, Message-ID: <69969198-6397-472c-ac73-609566daee5a@atl1s07mta2081.xt.local>, mail_id: czLde8GqdWpn, Hits: -0.089, size: 108816, queued_as: DE6F2148568, 5545 ms
Jul 19 15:01:09 mail postfix/qmgr[3643]: DE6F2148568: from=<bounce-344_HTML-18444363-1683099-7207065-3001@bounce.mail.arbonnemail.com>, size=109246, nrcpt=1 (queue active)
Jul 19 15:01:09 mail dovecot: lmtp(29810): Connect from local
Jul 19 15:01:09 mail postfix/smtp[29780]: CF302148566: to=<myuser@mydomain.com>, orig_to=<Skincare@mydomain.com>, relay=127.0.0.1[127.0.0.1]:10024, delay=6.4, delays=0.85/0/0/5.5, dsn=2.0.0, status=sent (250 2.0.0 from MTA(smtp:[127.0.0.1]:10025): 250 2.0.0 Ok: queued as DE6F2148568)
Jul 19 15:01:09 mail postfix/qmgr[3643]: CF302148566: removed
Jul 19 15:01:10 mail dovecot: lmtp(myuser@mydomain.com)<29810><iDgSN/XL9WBydAAAJmGFMg>: sieve: msgid=<69969198-6397-472c-ac73-609566daee5a@atl1s07mta2081.xt.local>: stored mail into mailbox 'INBOX'
Jul 19 15:01:10 mail postfix/lmtp[29809]: DE6F2148568: to=<myuser@mydomain.com>, relay=mail.mydomain.com[/var/run/dovecot/lmtp], delay=0.09, delays=0.01/0/0/0.08, dsn=2.0.0, status=sent (250 2.0.0 <myuser@mydomain.com> iDgSN/XL9WBydAAAJmGFMg Saved)
Jul 19 15:01:10 mail postfix/qmgr[3643]: DE6F2148568: removed
Jul 19 15:01:10 mail dovecot: lmtp(29810): Disconnect from local: Client has quit the connection (state=READY)
```

### PR DESCRIPTION
**Pull request by georglauterbach**
# Description

See https://github.com/docker-mailserver/docker-mailserver/issues/2091#issuecomment-943580324.

Fixes #2091

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by georglauterbach** [comment]
@docker-mailserver/maintainers tests pass, all seems well. Not sure why my local systems reported errors. Do I need extra software to run tests these days (like `jq`)?

**Comment by georglauterbach** [comment]
@casperklein all of your suggestions are valid, I just didn't have the time yet. Will tend to it ASAP.

**Comment by casperklein** [comment]
IIRC there was a bat test in another issue, which could reproduce the problem (cannot find it right now 😞). It might be a good idea to add the test for this?

**Comment by andrewlow** [comment]
FWIW - there are some existing tests for aliases in the testing framework.

I was trying to create something specific to detect the dovecot error saying "I don't know who this alias is" - then later once the alias is mapped - the mail is delivered.

**Review by polarathene — DISMISSED** [review]
LGTM :+1:

**Comment by wernerfred** [comment]
> * [Original request in Sep 2018](https://github.com/docker-mailserver/docker-mailserver/issues/1028), in 2019 maintainer @wernerfred expressed interest in the quota support. Presumably he still uses it?

yep, still using it!

> I can understand the value in the feature, but as this failure would have been silent and our workaround not ideal, probably best to make it opt-in as a breaking change release note

Definitly a breaking change imho

> So it's always been default since the introduction in Apr 2020 (v8 release).
I'm fine with making it opt-in instead. That should be handled via separate PR however.

But making it opt-in doesn't solve the problem, does it? The people that enable quotas would still face this behavior? Or am i missunderstanding something (admitting not reading all 55 comments and linked issues/pr comments in detail)?

**Comment by andrewlow** [comment]
>     So it's always been default since the introduction in Apr 2020 (v8 release).
>     I'm fine with making it opt-in instead. That should be handled via separate PR however.
> 
> But making it opt-in doesn't solve the problem, does it? The people that enable quotas would still face this behavior? Or am i missunderstanding something (admitting not reading all 55 comments and linked issues/pr comments in detail)?

The "problem" here is extra entries in the log files - where for every alias, you will see an error indicating that dovecot cannot find the alias user when it is trying to check the quota.

This problem is invisible, until you enable the optional logwatch reporting https://docker-mailserver.github.io/docker-mailserver/edge/config/environment/#logwatch_interval

Logwatch is detecting these log lines about the alias as an error, and reporting. It makes the logwatch report quite difficult to read if you have a lot of traffic to aliases.

**Comment by wernerfred** [comment]
Thanks for clarifying!
So the "problem" is just a stylistic/extra log line problem. But in fact i am right with the assumption that disabling the env by default doesn't solve the problem and users with enabled env will still "see" those log lines.
If this is the only outcome of changing the default behavior of the setting i would suggest keeping it as it is now (to avoid a breaking change) but to mention this issue in the documentation and how to solve it (disabling the quota env)
Wouldn't that be the "better" solution ?

**Comment by polarathene** [comment]
> (admitting not reading all 55 comments and linked issues/pr comments in detail)

@wernerfred this comment should cover everything you need to know, if not no worries let me know and I'll clarify/expand further :+1: 

---

> So the "problem" is just a stylistic/extra log line problem.

No it is not. This affects anyone using aliases and the default quota feature for Dovecot.

It risks making the `docker-mailserver` instance vulnerable to [backscatter](https://en.wikipedia.org/wiki/Backscatter_%28email%29), which allows spammers to forge a return address, bypass the Postfix `quota-status` policy check to reject mail at the `smtpd_recipient_restrictions` stage, delivering the mail to Dovecot which if quota has exceed, bounces it back to the return address.

**That can get you blacklisted**.

The quota feature itself is useful to enforce storage limitations for mail, or per user to avoid any abuse that could fill up the mail server storage, and thus no longer having space for others to store received mail.

If you don't need that, disabling quota support avoids the backscatter issue with aliases (still requires quota to actually be exceeded though).

The log message itself is only because Postfix is checking with Dovecot with the alias account in the `RCPT TO` envelope recipient address before reaching a stage that alias expansion happens (_once mail passes criteria to accept and hand off to the `cleanup` process which will add it to the "incoming queue"_).

---

> with the assumption that disabling the env by default doesn't solve the problem and users with enabled env will still "see" those log lines.

With this PR they would not see the log lines. The workaround is to add the alias account as a dummy account to Dovecot which shares the same mailbox of the real account with a quota.

The workaround is probably a bit naive though, in that:
- Aliases can map to other aliases.
- Not only can aliases nest/chain many times, but you can have one alias expand into multiple addresses, not just a 1-to-1 mapping.

`postmap -q <alias> <table>` can be used to query the alias (_for `docker-mailserver` with a plain-text file alias table and if not using the regexp alias table then we can do the same ourselves manually_). Unfortunately there doesn't appear to be support from the Postfix utilities to resolve the alias expansion fully.

Due to that, this workaround is naive in that it will only resolve simple alias mappings that map an alias to a single real account directly AFAIK. The proper fix would be to not have Dovecot care about these dummy accounts at all, and write a policy service for Postfix to call that resolves the alias fully and checks any mapped accounts to the Dovecot `quota-status` service, this would be a proxy service and requires more effort to implement.

The quota support is also not implemented by `docker-mailserver` for LDAP users, making it opt-in for non-LDAP would ensure a more consistent base setup and make troubleshooting potentially related issues a bit easier when we know the user reports opt-in via `ENABLE_QUOTAS=1`.

Alternatively we could implement quota support for LDAP too, but I don't think anyone here is interested in tackling that :sweat_smile: 

---

## Summary

### If you use quotas

#### The problem

- This log will appear with aliases due to no equivalent account on Dovecot as Postfix does not expand aliases prior to accepting mail and this log entry occurs when validating the `RCPT TO` against a quota policy check to Dovecot so that the mail can be rejected early and avoid enabling spammers via backscatter, which can get you blacklisted.
- The `quota-status` policy check will only reject when Dovecot knows it will reject for exceeding quota, if an account is not known, this will not result in rejection, so aliases will bypass the check and then possibly get rejected if quota exceeded.
- You do not want to reject when accounts aren't known to Dovecot during this check however as that'd always result in no delivery from that alias.

#### The workaround fix

- The PR provides a workaround by creating dummy accounts for Dovecot to be aware of the alias accounts.
- These are only added if the alias has a simple 1-to-1 mapping with a real account, not to multiple accounts or additional aliases, those will still run into the problem.

### If you don't use quotas

- The policy check for quotas isn't performed, so no problem.
- Your only problem without quotas is trust that you won't have storage issues that could be abused and negatively impact other users.
- LDAP users don't have quota support implemented by `docker-mailserver`.

---

## TL;DR

- The log is about an actual problem and shouldn't be dismissed, it can get you blacklisted for backscatter abuse from spammers.
- The fix is to not use quotas, or use the PR workaround that creates dummy Dovecot accounts for simple alias mappings.
- A better fix could be implemented but is more effort (_proxying the policy service to resolve aliases fully and query all resolved addresses to the quota policy service_).
- As `docker-mailserver` does not have quota support for LDAP users, it's probably better to make it opt-in for a more consistent base setup that assists with troubleshooting.

**Comment by polarathene** [comment]
> From a user perspective, the complimentary ENV file defines the defaults.

@andrewlow the complimentary ENV file is it's own set of defaults. Not all users add that. I don't expect users to know about internals, but for public facing ENV defaults, [our ENV documentation](https://docker-mailserver.github.io/docker-mailserver/edge/config/environment/) should be a more reliable source for users to know what is default. I would like to better automate genating that ENV file from our docs to keep them in sync though.

The source you provided was about updating that ENV file, while the actual quota support had an implicit default not "documented" in that ENV file for quite some time prior... an entire year.

> The "problem" here is extra entries in the log files - where for every alias, you will see an error indicating that dovecot cannot find the alias user when it is trying to check the quota.
> 
> This problem is invisible, until you enable the optional logwatch reporting

It becomes a real problem when quota is exceeded. See my previous comment for details.

---

> i would suggest keeping it as it is now (to avoid a breaking change) but to mention this issue in the documentation and how to solve it (disabling the quota env)
> Wouldn't that be the "better" solution ?

@wernerfred 

The issue probably still needs to be documented, especially the caveat for the workaround proposed.

A breaking change doesn't have to be introduced straight away, I've asked for a separate PR for any default ENV change. This way it can be merged at a later date as there are other changes inbound unrelated to this PR that are likely to warrant a major version bump.

File: target/scripts/startup/setup-stack.sh [hunk 1]
```diff
@@ -345,28 +346,98 @@ function _setup_dovecot_local_user
...
+      # ${REAL_FQUN} is a user's fully-qualified username
+      local ALIAS REAL_FQUN
+      while read -r ALIAS REAL_FQUN
+      do
+        # ignore comments
+        [[ ${ALIAS} == \#* ]] && continue
+
+        # alias is assumed to not be a proper e-mail
+        # these aliases do not need to be added to Dovecot's userdb
+        [[ ! ${ALIAS} == *@* ]] && continue
```
[the remark this answers is not in the document]
Comment (polarathene) [reply target/scripts/startup/setup-stack.sh:394]: This problem only applies to inbound external mail AFAIK.

I would need to check if `smtpd_recipient_restrictions` apply or not when a local service in the container sends mail to Postfix to deliver to Dovecot.

Even if it does, we have some rules earlier in these restrictions such as `permit_mynetworks` which will approve accepting the mail earlier. This prevents checking the quota policy against Dovecot. That'd probably not be great usually, but when the mail originates from you as the sender, then the backscatter concern of bouncing mail back to yourself probably doesn't apply?

I am not familiar with `amavis` other than it having something to do with spam handling. If it's not a concern, and the alias maps to an account that Dovecot receives mail for, and that quota is exceeded, then Dovecot bounces the mail (_which from descriptions I read, is repeated multiple times :man_shrugging: _)

File: target/scripts/startup/setup-stack.sh [hunk 2]
```diff
@@ -345,28 +346,98 @@ function _setup_dovecot_local_user
...
+
+        # clear possibly already filled arrays
+        unset REAL_ACC USER_QUOTA
+        declare -a REAL_ACC USER_QUOTA
+
+        local REAL_USERNAME REAL_DOMAINNAME
+        REAL_USERNAME=$(cut -d @ -f1 <<< "${REAL_FQUN}")
+        REAL_DOMAINNAME=$(cut -d @ -f2 <<< "${REAL_FQUN}")
```
[the remark this answers is not in the document]
Comment (polarathene) [reply target/scripts/startup/setup-stack.sh:402]: As a shell noob I find with the space easier to read, especially if the value is quoted :sweat_smile: 

I usually just look up the docs for the command and eventually get the gist of it, as `@` can be used as syntax for array stuff, someone like myself might get a bit confused, more so when it's not separated from `-d` :grimacing:

File: target/scripts/startup/setup-stack.sh [hunk 3]
```diff
@@ -345,28 +346,98 @@ function _setup_dovecot_local_user
...
+        #
+        # | position | content                                       |
+        # |        0 | real account name (e-mail address) == ${FQUN} |
+        # |        1 | password hash                                 |
+        # |        2 | optional user attributes                      |
+        IFS='|' read -r -a REAL_ACC < <(grep "${REAL_FQUN}" /tmp/docker-mailserver/postfix-accounts.cf)
+
+        if [[ -z ${REAL_ACC[1]} ]]
+        then
+          _notify 'err' "Could not acquire password hash of user '${REAL_FQUN}' while adding alias '${ALIAS}'"
```
[the remark this answers is not in the document]
Comment (polarathene) [reply target/scripts/startup/setup-stack.sh:422]: What is the difference between `err` and `fatal` if not killing `docker-mailserver`?

This will be run during `check-for-changes.sh` as well. If `fatal` would imply bringing down the container it probably shouldn't happen during change events :sweat_smile: (then again, that's probably going to be the case when it starts calling the `SSL_TYPE` handling methods which call `dms_panic` for some failures.

[the remark this answers is not in the document]
Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:422]: Wrong word choice. I meant something fatal like dms_panic.

File: target/scripts/startup/setup-stack.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:402]: Or the other way: add space to -f

**Comment by wernerfred** [comment]
> @wernerfred this comment should cover everything you need to know, if not no worries let me know and I'll clarify/expand further 👍

Insane, thank you!

So what i am not understanding right now is how `POSTFIX_MAILBOX_SIZE_LIMIT` is involved in this issue here. Does setting this ENV (and therefore limiting everys mailbox size) prevent exploiting this issue?

**Comment by polarathene** [comment]
> So what i am not understanding right now is how `POSTFIX_MAILBOX_SIZE_LIMIT` is involved in this issue here. Does setting this ENV (and therefore limiting everys mailbox size) prevent exploiting this issue?

Where are you referring to this from? I recall that being part of the quota feature PR, but not this one. [Postfix `main.cf:mailbox_size_limit`](http://www.postfix.org/postconf.5.html#mailbox_size_limit) is what you're referring to? I don't think this has anything to do with Dovecot Quota, but is similar for the Postfix `local` delivery agent, which shouldn't need to go through any recipient restrictions to be accepted by Postfix AFAIK.

I've not looked into that personally so I don't quite know what happens there, presumably it relays over LMTP to Dovecot, if so then either it bypasses the quota check and could be an issue, or by relaying via LMTP it may go through the proper checks again on the recipient, however by this time Postfix probably already expanded the alias via `local` delivery first.

File: target/scripts/startup/setup-stack.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:394]: Thank you for clarification. FYI, I use the amavis alias because of [this](https://github.com/docker-mailserver/docker-mailserver/issues/1224#issuecomment-552575494).

[the remark this answers is not in the document]
Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh:394]: As long as postmaster@example.com is not a _real_ address (in `postfix-accounts.cf`, this PR will not change / add anything as (as @polarathene has pointed out), only 1-to-1 aliases (alias -> real user) are expandend. Should all be fine here then?

File: target/scripts/startup/setup-stack.sh [hunk 4]
```diff
@@ -345,28 +346,98 @@ function _setup_dovecot_local_user
...
+      do
+        # ignore comments
+        [[ ${ALIAS} == \#* ]] && continue
+
+        # alias is assumed to not be a proper e-mail
+        # these aliases do not need to be added to Dovecot's userdb
+        [[ ! ${ALIAS} == *@* ]] && continue
+
+        # clear possibly already filled arrays
+        unset REAL_ACC USER_QUOTA
```
[the remark this answers is not in the document]
Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh:397]: I'm not too sure about this. `read` does not clear the array, but just appends from the beginning.

In our case, this means always having optional user attributes if one user has them for all the users that come afterwards (in the `while` loop) and the password hash test being useless really. I would not remove it as it does not harm, and **I also think it actually is required**.

File: target/scripts/startup/setup-stack.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh:422]: Hmm, I see. On the one hand, one could argue that shutting was too hard here, but then, this error also indicates that something is wrong in `/tmp/docker-mailserver/postfix-accounts.cf`, which is fatal, I agree. I will tend to it.

File: target/scripts/startup/setup-stack.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:394]: I thought the whole point of this PR is to avoid some nasty error messages? So in any case, this PR should not harm in any way. Or do I miss something fundamental at the moment?

postmaster@example.com is a real address (in postfix-accounts.cf). Why it shouldn't?

File: target/scripts/startup/setup-stack.sh [hunk 5]
```diff
@@ -342,31 +343,98 @@ function _setup_dovecot_local_user
       if [[ -f /tmp/docker-mailserver/dovecot-quotas.cf ]]
       then
         declare -a USER_QUOTA
-        IFS=':' ; read -r -a USER_QUOTA < <(grep "${USER}@${DOMAIN}:" -i /tmp/docker-mailserver/dovecot-quotas.cf)
+        IFS=':' read -r -a USER_QUOTA < <(grep "${USER}@${DOMAIN}:" -i /tmp/docker-mailserver/dovecot-quotas.cf)
         unset IFS
```
Comment (casperklein) [inline target/scripts/startup/setup-stack.sh:347]: IFS always have a default values.
We don't change this value generally, just for the specific line 346.
If we unset IFS, then IFS is empty.
For example, this would be equally to putting "unset IFS" at the start of all scripts - pointless.

---

Before this change, there was a `;` after `IFS=':'`. This way, the default value was changed to `:`. Even then, restoring the old IFS value would be better instead of completely unseting it.

```suggestion
```

File: target/scripts/startup/setup-stack.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh:394]: > [postmaster@example.com](mailto:postmaster@example.com) is a real address (in postfix-accounts.cf). Why it shouldn't?

This was just an assumption (in this case, an entry will be added to Dovecot's userdb). This PR (should) does not do harm in any way, of course.

File: target/scripts/startup/setup-stack.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:397]: I think, I lost a bit the overview. I didn't realize, this was part of a loop.

File: target/scripts/startup/setup-stack.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:393]: Got it. But then, when someone have something like this in `postfix-virtual.cf`, this would not add an entry to dovecots userdb?

`amavis not-a-real-account-just-another-alias@example.com` in `postfix-virtual.cf`

PS: I don't know, if it's valid to point an alias to an other alias.

[the remark this answers is not in the document]
Comment (polarathene) [reply target/scripts/startup/setup-stack.sh:393]: > That's why I have something like this: `amavis postmaster@example.com` in `postfix-virtual.cf`.

I believe that should be in `postfix-aliases.cf` (_parameter `main.cf:alias_maps`_) like the comment you linked to mentions with aliasing the `amavis` account to `root` (which can alias again to `postmaster@<FQDN>`.

`postfix-virtual.cf` (_parameter `main.cf:virtual_alias_map` with generated file `/etc/postfix/virtual`_) and `postfix-accounts.cf` (_parameter `main.cf:virtual_mailbox_maps` with generated file `/etc/postfix/vmailbox`_) provide unique domains for the generated file `/etc/postfix/vhost` (_parameter `main.cf:virtual_mailbox_domains`).

Postfix should receive mail from local account without any FQDN, eg from `amavis` and match it to an alias in `hash:/etc/aliases` (_`/etc/aliases` is plain-text, hash database generated by `postalias` is `/etc/aliases.db` that Postfix uses, values are sourced from `postfix-aliases.cf` populating `/etc/aliases` at startup or change event_). Mappings in `/etc/aliases` lack an FQDN, but Postfix will under some conditions receive mail and extract the local part (eg left side of `@`) from incoming mail recipient and lookup these `/etc/aliases` if no explicit match was found in virtual aliases.

Since `postfix-aliases.cf` (_unix user aliases_) is purely meant for these aliases, while `postfix-accounts.cf` (_virtual mailboxes, storage not tied to unix user_) and `postfix-virtual.cf` (_virtual aliases, no unix user_) are handled by us differently (_contributing to vhost domains_), I think you would place `amavis` and the like in `postfix-aliases.cf` :sweat_smile: (don't worry, I'm probably as confused as you are right now :stuck_out_tongue: )

---

> Got it. But then, when someone have something like this in `postfix-virtual.cf`, this would not add an entry to dovecots userdb?

[Postfix address rewriting docs on virtual aliases](http://www.postfix.org/ADDRESS_REWRITING_README.html#virtual) shows an example of a 1-to-1 mapping in `postfix-virtual.cf` that's got no associated FQDN, so even I'm confused about it vs the `postfix-alias.cf` kind. This is perfectly valid as you point out..

Dovecot AFAIK doesn't care what the user name is, eg `amavis` or `amavis@example.com`? What I believe happens here is Postfix will append `main.cf:myorigin` or if not available another fallback domain it's aware of, and then deliver that to Dovecot to store in `/var/mail`.

In the case of Amavis, it should only be mail to and from your server, if Dovecot can't be configured to ignore quota for system user accounts (eg: `postmaster@example.com`) you could make the quota really high. In regards to the issue this PR is solving, there are "permit" rules that can come before the quota policy check, eg `mynetworks` and the one for SASL authenticated is probably safe and both of these are configured to come before the quota check, thus Postfix will accept and if Dovecot rejects, it bounces but if the accounts that sent it aren't external, the backscatter blacklisting concern shouldn't apply?

---

> PS: I don't know, if it's valid to point an alias to an other alias.

I have read of an `all` alias that expands to many users, some which I guess could be aliases. 1-to-1 alias that maps to another alias seems a bit silly, but our docs have an example of working around `SPOOF_PROTECTION` issue for [aliasing to gmail via an alias to alias mapping which then maps the 2nd level alias to the external account](https://github.com/docker-mailserver/docker-mailserver/blob/fb72f3ad52769ad470cd1f7fc36f76b53853f3e8/docs/content/examples/tutorials/basic-installation.md?plain=1#L130-L135).

---

Probably good to eventually investigate further the above for definitive answers about account/alias handling and document that somewhere, but I don't have time to allocate to that right now :grimacing:

[the remark this answers is not in the document]
Comment (polarathene) [reply target/scripts/startup/setup-stack.sh:393]: Just to clarify, the log issue with Postfix asking Dovecot for quota status of an alias that Dovecot doesn't know about is always the `RCPT TO` (envelope recipient) address the email was originally sent to.

Regardless of nested chains, Postfix does not expand that aliases at this point (_which could be N levels deep or expand to multiple addresses to send to_). In the docs mention in my previous comment, that means the gmail account with double aliases won't be added to Dovecot, not that it matters since that's aliasing to an external account anyway.

If it ever turns out to be an issue, we can probably regard it as niche and unsupported. Our responsibility would just be to make the caveat with quota support with aliases visible in our docs, and that we only assist with direct aliasing to a single real mailbox account, any fancy aliasing beyond that puts the user as risk of backscatter if they have quotas enabled.

File: target/scripts/startup/setup-stack.sh (hunk 5 above)
Comment (polarathene) [reply target/scripts/startup/setup-stack.sh]: `unset IFS` seems fine for bash to reset to default value, it will not make IFS empty:

- https://unix.stackexchange.com/questions/97481/ifs-null-is-not-the-same-as-unset-ifs/97492#97492
- https://stackoverflow.com/questions/62470380/why-cant-reset-ifs-variable/62470888#62470888

Comment (casperklein) [reply target/scripts/startup/setup-stack.sh]: > We also have unset IFS present in check-for-changes.sh, helper-functions.sh and setup-stack.sh already.

Removing the `;` before `read` like here would also allow to remove `unset IFS` there.

Comment (georglauterbach) [reply target/scripts/startup/setup-stack.sh]: @casperklein is right, removing the `unset IFS` here is a good addition. I'd just like to point out that this PR is ultimately not about removing `unset`s.

**Review by polarathene — APPROVED** [review]

File: target/scripts/startup/setup-stack.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:393]: > don't worry, I'm probably as confused as you are right now 

😆 

Thank you for you detailed answer!

**Review by casperklein — APPROVED** [review]

**Comment by polarathene** [comment]
This happened again..

I wrote the commit message for the PR while tests were running via the "enable auto-squash merge" button. It did nothing on confirmation, when I reloaded the page, there was an error up top in red about it failing. The action had then changed to "squash merge" while the tests continued to run.

I'm not familiar with the setup, so perhaps that's expected, and only the lint test blocks merging once approvals are given?

**Comment by georglauterbach** [comment]
> I'm not familiar with the setup, so perhaps that's expected, and only the lint test blocks merging once approvals are given?

This is indeed the case :)
